File: Problem/By section/Binary Search/Problem 278/problem_278.cpp
#include <bits/stdc++.h>

using namespace std;

bool isBadVersion(int version);

class Solution {
public:
    // Time: O(log(n)), Space: O(1)
    int firstBadVersion(int n) {
        int left = 1, right = n;
        int mid;
        while(left <= right) {
            mid = left + (right - left) / 2;
            if(isBadVersion(mid)) {
                right = mid - 1;
            }else if(!isBadVersion(mid)) {
                left = mid + 1;
            }
        }

        if(left > n) return -1;

        return left;
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}

File: Problem/By section/Binary Search/Problem 278/problem_278.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

class Solution extends VersionControl {
    // Time: O(log(n)), Space: O(1)
    public int firstBadVersion(int n) {
        if(n <= 0) return 1;
        
        return bin_search(n, 0, n);
        
    }
    
    public int bin_search(int n, int l, int r){
        int middle = l + (r - l) / 2;
        
        if(l == r){
            return l;
        }else if(isBadVersion(middle)){
            // if we know the middle one is bad, we know that the first bad version can be in [l, l + 1,..., middle]
            return bin_search(n, l, middle);
        }else{
            // else then the bad vision can only happen in [middle + 1,..., r]
            return bin_search(n, middle + 1, r);
        }
    }
}